messaging/rust/src/messaging.rs

```diff
@@ -0,0 +1,296 @@
+};
+
+pub const SMITHY_VERSION: &str = "1.0";
+
+/// A message to be published
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct PubMessage {
+    /// The message payload
+    #[serde(with = "serde_bytes")]
+    #[serde(default)]
+    pub body: Vec<u8>,
+    /// An optional topic on which the reply should be sent.
+    #[serde(rename = "replyTo")]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub reply_to: Option<String>,
+    /// The subject, or topic, of the message
+    #[serde(default)]
+    pub subject: String,
+}
+
+/// Reply received from a Request operation
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct ReplyMessage {
+    /// The message payload
+    #[serde(with = "serde_bytes")]
+    #[serde(default)]
+    pub body: Vec<u8>,
+    /// An optional topic on which the reply should be sent.
+    #[serde(rename = "replyTo")]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub reply_to: Option<String>,
+    /// The subject, or topic, of the message
+    #[serde(default)]
+    pub subject: String,
+}
+
+/// Message sent as part of a request, with timeout
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct RequestMessage {
+    /// The message payload
+    #[serde(with = "serde_bytes")]
+    #[serde(default)]
+    pub body: Vec<u8>,
+    /// The subject, or topic, of the message
+    #[serde(default)]
+    pub subject: String,
+    /// A timeout, in milliseconds
+    #[serde(rename = "timeoutMs")]
+    pub timeout_ms: u32,
+}
+
+/// Message received as part of a subscription
+#[derive(Clone, Debug, Default, Deserialize, Eq, PartialEq, Serialize)]
+pub struct SubMessage {
+    /// The message payload
+    #[serde(with = "serde_bytes")]
+    #[serde(default)]
+    pub body: Vec<u8>,
+    /// An optional topic on which the reply should be sent.
+    #[serde(rename = "replyTo")]
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub reply_to: Option<String>,
+    /// The subject, or topic, of the message
+    #[serde(default)]
+    pub subject: String,
+}
+
+/// The MessageSubscriber interface describes
+/// an actor interface that receives messages
+/// sent by the Messaging provider
+/// wasmbus.contractId: wasmcloud:messaging
+/// wasmbus.actorReceive
+#[async_trait]
+pub trait MessageSubscriber {
+    /// returns the capability contract id for this interface
+    fn contract_id() -> &'static str {
+        "wasmcloud:messaging"
+    }
+    /// subscription handler
+    async fn handle_message(&self, ctx: &Context, arg: &SubMessage) -> RpcResult<()>;
+}
+
+/// MessageSubscriberReceiver receives messages defined in the MessageSubscriber service trait
+/// The MessageSubscriber interface describes
+/// an actor interface that receives messages
+/// sent by the Messaging provider
+#[async_trait]
+pub trait MessageSubscriberReceiver: MessageDispatch + MessageSubscriber {
```

Comment:
Can this be pub(crate), otherwise people will see the message subscriber and message subscriber receiver traits and be confused as to which one they implement.

If it can't be crate only we can doc(hidden) it to keep it from showing up in user facing docs.

Reply:
Cool. I was just primarily interested in avoiding confusion on the part of the developer.

Reply:
will fix in codegen crate